Mixed integer rotation constraint

hongkai-dai opened the pull request:
This PR cleans up the mixed integer relaxation on SO(3) constraint. The key features in this PR includes

1. Add a new class `MixedIntegerRotationConstraintGenerator`, with the following features
   1. Now we can consider three relaxation approaches, by using bilinear McCormick envelope, or by considering the intersection between boxes and sphere surface, or by combining the former two approaches.
   2. For each approach, we can add either linear number of binary variables, or logarithmic number of binary variables (w.r.t the number of intervals in [-1, 1]).
   3. When considering the box-sphere intersection, we need to solve an SOCP. The result of the SOCP is the same, if we do not change the number of intervals. But previously to add mixed-integer relaxation to several rotation matrices (for example in IK case, we need to relax the rotation for each link), for each of that matrix we need to solve the same SOCP again. Now the result of that SOCP is stored inside `MixedIntegerRotationConstraintGenerator`.

2. Create new files `mixed_integer_rotation_constraint.h/cc`. The code previously in `rotation_constraint.h/cc` that involves mixed integer variables, are all moved to the new files.
3. The function previously called `AddRotationMatrixMcCormickEnvelopeMilpConstraints` is renamed as `AddRotationMatrixBoxSphereIntersectionMilpConstraints`, and now it lives in `mixed_integer_rotation_constraint.h`. McCormick envelope is to relax the non-convex constraint, using sub-gradient in each interval. I do not think that is exactly what we did in that function.

@gizatt and @RussTedrake

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8748)
<!-- Reviewable:end -->

hongkai-dai commented:
@drake-jenkins-bot mac-highsierra-clang-bazel-experimental-everything please
@drake-jenkins-bot mac-sierra-clang-bazel-experimental-everything please
@drake-jenkins-bot mac-highsierra-clang-bazel-experimental-everything-debug please

---

Review status: 0 of 21 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBYGja05jfhzEtrsynh:b20vxnl)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
-(status: do not review) 
+@gizatt for feature review please.

---

Review status: 0 of 22 files reviewed at latest revision, all discussions resolved.

---

*[solvers/mixed_integer_rotation_constraint.h, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBZ4qFN56yQw7cvPUWs:-LBZ4qFN56yQw7cvPUWt:b-q3kisx) ([raw file](https://github.com/robotlocomotion/drake/blob/dd14d4d838cf06b6f8dcd62fed19ef94cfecb45f/solvers/mixed_integer_rotation_constraint.h#L1)):*
> ```objc
> #pragma once
> ```

To reviewers: this header file is new code, please review it, thanks!

---

*[solvers/mixed_integer_rotation_constraint.cc, line 349 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBZ3QcR0iPEd7-ljr4O:-LBZ3QcR0iPEd7-ljr4P:b-guea0r) ([raw file](https://github.com/robotlocomotion/drake/blob/dd14d4d838cf06b6f8dcd62fed19ef94cfecb45f/solvers/mixed_integer_rotation_constraint.cc#L349)):*
> ```cpp
>       break;
>     }
>     case IntervalBinning::kLinear: {
> ```

To reviewers: the code in this `case IntervalBinning::kLinear` part is new code, please review it, thanks

---

*[solvers/mixed_integer_rotation_constraint.cc, line 859 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBZ3qXa8piDuYevD6EK:-LBZ3qXb4dQhpH3rhqEP:b-ziatrw) ([raw file](https://github.com/robotlocomotion/drake/blob/dd14d4d838cf06b6f8dcd62fed19ef94cfecb45f/solvers/mixed_integer_rotation_constraint.cc#L859)):*
> ```cpp
> }
> 
> MixedIntegerRotationConstraintGenerator::
> ```

To reviewers: the code inside this constructor function is new code, please review it, thanks!

---

*[solvers/mixed_integer_rotation_constraint.cc, line 887 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBZ40xTBsXLa2CV_Yg9:-LBZ40xTBsXLa2CV_YgA:b1llg10) ([raw file](https://github.com/robotlocomotion/drake/blob/dd14d4d838cf06b6f8dcd62fed19ef94cfecb45f/solvers/mixed_integer_rotation_constraint.cc#L887)):*
> ```cpp
> 
> MixedIntegerRotationConstraintGenerator::ReturnType
> MixedIntegerRotationConstraintGenerator::AddToProgram(
> ```

To reviewers: the code inside `AddToProgram` is new code, please review it, thanks!

---

*[solvers/mixed_integer_rotation_constraint_internal.h, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBZ4gX7CetjEfDDgQqJ:-LBZ4gX7CetjEfDDgQqK:b-2q2ka) ([raw file](https://github.com/robotlocomotion/drake/blob/dd14d4d838cf06b6f8dcd62fed19ef94cfecb45f/solvers/mixed_integer_rotation_constraint_internal.h#L1)):*
> ```objc
> #pragma once
> ```

To reviewers: this header file is just renaming the existing file"rotation_constraint_internal.h", you can ignore reviewing it.

---

*[solvers/mixed_integer_rotation_constraint_internal.cc, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBZ4DDoETWLaPr0YgmT:-LBZ4DDoETWLaPr0YgmU:bruhhak) ([raw file](https://github.com/robotlocomotion/drake/blob/dd14d4d838cf06b6f8dcd62fed19ef94cfecb45f/solvers/mixed_integer_rotation_constraint_internal.cc#L1)):*
> ```cpp
> #include "drake/solvers/mixed_integer_rotation_constraint_internal.h"
> ```

To reviewer: this diff just moves some code from `rotation_constraint.cc` to this new file, you can ignore reviewing it.

---

*[solvers/test/mixed_integer_rotation_constraint_test.cc, line 187 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBZ5I43BxxYfK1usIO3:-LBZ5I43BxxYfK1usIO4:bio54al) ([raw file](https://github.com/robotlocomotion/drake/blob/dd14d4d838cf06b6f8dcd62fed19ef94cfecb45f/solvers/test/mixed_integer_rotation_constraint_test.cc#L187)):*
> ```cpp
> };
> 
> TEST_P(TestMixedIntegerRotationConstraintGenerator, TestConstructor) {
> ```

To reviewers: all code in this file, except this `TestConstructor` and the next `TestBinaryAssignment` functions, are moved from `rotation_constraint_test`. You may ignore reviewing all pieces except these two functions. Thanks!

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBZArJ23LWTsY2H_lzp:bfjz0u)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 22 files reviewed at latest revision, 7 unresolved discussions.

---

*[solvers/mixed_integer_rotation_constraint.cc, line 1 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBZHBwD90qQkF03XKs6:-LBZHBwD90qQkF03XKs7:b-vv1xj5) ([raw file](https://github.com/robotlocomotion/drake/blob/482137d3668ffdda95bba0b7f056a15be61d74a6/solvers/mixed_integer_rotation_constraint.cc#L1)):*
> ```cpp
> #include "drake/solvers/mixed_integer_rotation_constraint.h"
> ```

To reviewers: most of this file is moved from `rotation_constraint.cc`. I marked the part with new code. You could ignore reviewing most of this file.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748)*
<!-- Sent from Reviewable.io -->

gizatt commented:
Checking in, have to move over to class.

---

Reviewed 16 of 23 files at r1, 1 of 1 files at r2, 1 of 2 files at r3.
Review status: 17 of 23 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[solvers/mixed_integer_optimization_util.h, line 44 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbAkJ15nq0ahtIIg5O:-LBbAkJ15nq0ahtIIg5P:b-fv7gga) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_optimization_util.h#L44)):*
> ```objc
>  *   λ(0) + ... + λ(n) = 1
>  *   ∀i. λ(i) ≥ 0
>  *   ∃ j ∈ {0, 1, ..., n-1}, s.t λ(i) = 0 if i ≠ j and i ≠ j + 1 
> ```

I think the "there exists... s.t." needs to be replaced with "for all"? (Or an equivalent rewording should happen?)

---

*[solvers/test/mixed_integer_rotation_constraint_limit_test.cc, line 94 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbBuofEJvggXQiRl0L:-LBbBuofEJvggXQiRl0M:b-sqt9l) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/test/mixed_integer_rotation_constraint_limit_test.cc#L94)):*
> ```cpp
>         min_distance = {{0.60302, 1.25649, 1.33283}};
>         break;
>       }
> ```

Keeping the default case could be useful in case more approaches are added later? Though it's only really useful to catch a test writer who modifies the test instantiation below but does not implement the corresponding case here.

---

*[solvers/test/mixed_integer_rotation_constraint_limit_test.cc, line 153 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbCqSr3w9gZWHFKC78:-LBbCqSr3w9gZWHFKC79:b8ve0ox) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/test/mixed_integer_rotation_constraint_limit_test.cc#L153)):*
> ```cpp
>                  kBoxSphereIntersection,
>              MixedIntegerRotationConstraintGenerator::Approach::
>                  kBilinearMcCormick}),
> ```

Does "both" ever get tested?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBbD2Dq6SUR2-lWynV5:b-uya3s9)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 17 of 23 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[solvers/mixed_integer_optimization_util.h, line 44 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbAkJ15nq0ahtIIg5O:-LBbH9641QvAWRmXgaoB:b-lhx07k) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_optimization_util.h#L44)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

I think the "there exists... s.t." needs to be replaced with "for all"? (Or an equivalent rewording should happen?)
</blockquote></details>

I think it should be `there exists one single j, such that when i is not j and i is not j+1, then lambda(i) has to be 0`. So I do not think we should replace with "for all".

---

*[solvers/test/mixed_integer_rotation_constraint_limit_test.cc, line 94 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbBuofEJvggXQiRl0L:-LBbHb265Mrw26BlkwmT:b-ivrwfn) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/test/mixed_integer_rotation_constraint_limit_test.cc#L94)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Keeping the default case could be useful in case more approaches are added later? Though it's only really useful to catch a test writer who modifies the test instantiation below but does not implement the corresponding case here.
</blockquote></details>

The compiler will throw an error, if there are more approaches added to the enum `MixedIntegerRotationConstraintGenerator::Approach`, but the switch statement here does not handle that enum case. So I think it is better not to use `default` here, as using `default` would not throw that compiler error, when the new approach is added.

---

*[solvers/test/mixed_integer_rotation_constraint_limit_test.cc, line 153 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbCqSr3w9gZWHFKC78:-LBbI0BI11RYcPdAKAXP:b-jxo46p) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/test/mixed_integer_rotation_constraint_limit_test.cc#L153)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Does "both" ever get tested?
</blockquote></details>

I removed it here in the instantiation, so as to avoid the test time getting too long (the "both" case takes about 20mins to run).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748)*
<!-- Sent from Reviewable.io -->

gizatt commented:
Reviewed 5 of 23 files at r1, 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[solvers/mixed_integer_optimization_util.h, line 44 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbAkJ15nq0ahtIIg5O:-LBbT7hO1Q5z5lbUMbhg:b3ock14) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_optimization_util.h#L44)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think it should be `there exists one single j, such that when i is not j and i is not j+1, then lambda(i) has to be 0`. So I do not think we should replace with "for all".
</blockquote></details>

Ah, I see what you mean. This does seem like a sort of backwards way of writing it, but I'm no longer convinced there's a dramatically better way of saying it.

---

*[solvers/mixed_integer_rotation_constraint.h, line 24 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbIjWx-wjGpHqfO8pS:-LBbIjWx-wjGpHqfO8pT:b-nwn8ig) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L24)):*
> ```objc
>  * (the range of each entry in R) into smaller intervals, and then relax the
>  * SO(3) constraint within each interval. We provide 3 approaches for relaxation
>  * 1. By replacing the bilinear product with a new variable, in the McCormick
> ```

BTW Maybe it would help to clarify that this is done for each bilinear product that appears in SO(3) constraints, and maybe additionally justify that implementing SO(3) constraints requires many bilinear constraints. Though, to be fair, the other constraint is complex too, so maybe referring to the paper is the best approach.

---

*[solvers/mixed_integer_rotation_constraint.h, line 26 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbHybn8b5o9lNHOlEJ:-LBbHybn8b5o9lNHOlEK:b-16qkyk) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L26)):*
> ```objc
>  * 1. By replacing the bilinear product with a new variable, in the McCormick
>  * envelope of the bilinear product w = x * y.
>  * 2. By consider the intersection region between axis-aligned boxes, and the
> ```

considering

---

*[solvers/mixed_integer_rotation_constraint.h, line 30 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbIEIQDjv8w-XL4_PO:-LBbIEIQDjv8w-XL4_PP:bfnghfc) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L30)):*
> ```objc
>  * 3. By combining the two approaches above.
>  * These three approaches give different relaxation of SO(3) constraint (the
>  * feasible sets to each relaxation are different), and different computation
> ```

to->for

---

*[solvers/mixed_integer_rotation_constraint.h, line 32 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbIHuz9LRSOMDZx139:-LBbIHuz9LRSOMDZx13A:bd0oix7) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L32)):*
> ```objc
>  * feasible sets to each relaxation are different), and different computation
>  * speed. The user can switch between the approaches to find the best fit for
>  * the problem.
> ```

their problem?

---

*[solvers/mixed_integer_rotation_constraint.h, line 34 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbITwo2ZQTMFWjMbdJ:-LBbITwo2ZQTMFWjMbdK:b4aoy05) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L34)):*
> ```objc
>  * the problem.
>  *
>  * @note If you have several rotation matrices, all need to be relaxed through
> ```

"several rotation matrices that all need to be..." flows a little better

---

*[solvers/mixed_integer_rotation_constraint.h, line 58 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbKlDU9b60LaLF0SN9:-LBbKlDU9b60LaLF0SNA:b-gi0pcq) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L58)):*
> ```objc
> 
>   struct ReturnType {
>     /**
> ```

BTW do you expect users to see that this is a 2D array of decision variable vectors from the function signature, or is it worth mentioning? Maybe at least worth mentioning that each decision variable vector is a vector of binary decision variables?

---

*[solvers/mixed_integer_rotation_constraint.h, line 61 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbKI5hBP3C0l3xP78n:-LBbKI5hBP3C0l3xP78o:bv3r352) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L61)):*
> ```objc
>      * B_[i][j] represents in which interval R(i, j) lies. If we use linear
>      * binning, then B_[i][j] is of length 2 * num_intervals_per_half_axis_.
>      * B_[i][j](k) = 1 => φ(k) ≤ R(i, j) ≤ φ(k + 1)
> ```

I didn't see phi is defined before this point

---

*[solvers/mixed_integer_rotation_constraint.h, line 72 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbL7I6A1bObcS7KBnl:-LBbL7I6A1bObcS7KBnm:b-qm7nvm) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L72)):*
> ```objc
>      * R(i, j) = φᵀ * λ_[i][j]. Notice that λ_[i][j] satisfies the special
>      * ordered set of type 2 (SOS2) constraint. Namely at most two entries in
>      * λ_[i][j] can be non-negative, and these two entries have to
> ```

non-negative -> positive? Because all but two elements of lambda are zero, which is non-negative, as written here? 

---

*[solvers/mixed_integer_rotation_constraint.h, line 103 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbL_qNENzfgYIXNQHe:-LBbL_qNENzfgYIXNQHf:bk7umn4) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L103)):*
> ```objc
>   /**
>    * Add the mixed-integer linear constraints to the optimization program, as
>    * an relaxation of SO(3) constraint on the rotation matrix `R`.
> ```

an -> a

---

*[solvers/mixed_integer_rotation_constraint.cc, line 899 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbQISeBtUQYX145WM3:-LBbQISeBtUQYX145WM4:bsnpyk1) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.cc#L899)):*
> ```cpp
>       // R(i, j) = φᵀ * λ[i][j]
>       prog->AddLinearConstraint(
>           R(i, j) -
> ```

BTW Why phrase as "R - phi.' lambda == 0" rather than "R = phi.' lambda"?

---

*[solvers/mixed_integer_rotation_constraint.cc, line 928 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbQwOb1X9Rbsz8UM-U:-LBbQwOb1X9Rbsz8UM-V:b-ratht6) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.cc#L928)):*
> ```cpp
>   // term in the McCormick envelope of w = x * y
>   if (approach_ == Approach::kBilinearMcCormick ||
>       approach_ == Approach::kBoth) {
> ```

BTW You could implement Approach with a bit field such that this check would be (approach_ & Approach::kBilinearMcCormick != 0), and you can request methods with an invocation like (Approach::Foo | Approach::Bar). But that's only cleaner if there are more than two approaches.

---

*[solvers/test/mixed_integer_rotation_constraint_limit_test.cc, line 153 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbCqSr3w9gZWHFKC78:-LBbS06m3lp1l9dm0Gvv:b3mpf77) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/test/mixed_integer_rotation_constraint_limit_test.cc#L153)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I removed it here in the instantiation, so as to avoid the test time getting too long (the "both" case takes about 20mins to run).
</blockquote></details>

Got it! Not much is lost there.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 20 of 22 files reviewed at latest revision, 19 unresolved discussions.

---

*[solvers/mixed_integer_rotation_constraint.h, line 24 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbIjWx-wjGpHqfO8pS:-LBbZ8bP9eaMZT_F_Y4d:b-ccndpx) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L24)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

BTW Maybe it would help to clarify that this is done for each bilinear product that appears in SO(3) constraints, and maybe additionally justify that implementing SO(3) constraints requires many bilinear constraints. Though, to be fair, the other constraint is complex too, so maybe referring to the paper is the best approach.
</blockquote></details>

Done. I added the formulation on the SO(3) constraint. Down below where I mention `IntervalBinning`, I explained how many binary variables will be introduced, by either linear or logarithmic binning.

---

*[solvers/mixed_integer_rotation_constraint.h, line 26 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbHybn8b5o9lNHOlEJ:-LBbZND39wvU30nztkDd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L26)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

considering
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 30 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbIEIQDjv8w-XL4_PO:-LBboE3uDwTp69EcU862:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L30)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

to->for
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 32 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbIHuz9LRSOMDZx139:-LBbZXC5AERURfmB6wK_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L32)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

their problem?
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 34 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbITwo2ZQTMFWjMbdJ:-LBbZbsa2MBKfBZ9NW4L:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L34)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

"several rotation matrices that all need to be..." flows a little better
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 58 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbKlDU9b60LaLF0SN9:-LBbZfs99jaxI5DuGkmE:b-fglp0y) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L58)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

BTW do you expect users to see that this is a 2D array of decision variable vectors from the function signature, or is it worth mentioning? Maybe at least worth mentioning that each decision variable vector is a vector of binary decision variables?
</blockquote></details>

I think the type of `B_` suggest that it is a 2D array of decision variable vectors? I added the comment that `B_ contains the new binary variables added to the program.`

---

*[solvers/mixed_integer_rotation_constraint.h, line 61 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbKI5hBP3C0l3xP78n:-LBb_BXCDAzPLKTdKAT5:bwi4ehm) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L61)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

I didn't see phi is defined before this point
</blockquote></details>

Good point, I added phi in the class documentation.

---

*[solvers/mixed_integer_rotation_constraint.h, line 72 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbL7I6A1bObcS7KBnl:-LBbnVD_4xKsFOgMd7Tj:b-376vyo) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L72)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

non-negative -> positive? Because all but two elements of lambda are zero, which is non-negative, as written here? 
</blockquote></details>

Done. Good catch!

---

*[solvers/mixed_integer_rotation_constraint.h, line 103 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbL_qNENzfgYIXNQHe:-LBbnb4PDGBzaWgQ4BuU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L103)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

an -> a
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.cc, line 899 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbQISeBtUQYX145WM3:-LBbne4m53YpNK7zJMnm:b-sufgrm) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.cc#L899)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

BTW Why phrase as "R - phi.' lambda == 0" rather than "R = phi.' lambda"?
</blockquote></details>

I think they are equivalent.

I change to use `AddLinearEqualityConstraint` which is faster to parse

---

*[solvers/mixed_integer_rotation_constraint.cc, line 928 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbQwOb1X9Rbsz8UM-U:-LBbo4osDb-zAXZ7LTNu:b-aenu4p) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.cc#L928)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

BTW You could implement Approach with a bit field such that this check would be (approach_ & Approach::kBilinearMcCormick != 0), and you can request methods with an invocation like (Approach::Foo | Approach::Bar). But that's only cleaner if there are more than two approaches.
</blockquote></details>

That is a good point! I agree it is more extensible, if we add more approaches. On the other hand, I have the concern that the user can misuse a wrong input, like `0b00` or `0b100`, as there is not a bitfield class in C++, while using enum class the user is guaranteed to use a right input.

Given this concern, do you know a good way in C++, to guarantee that the user chooses a valid input using bit field?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748)*
<!-- Sent from Reviewable.io -->

gizatt commented:
With current changes, and assuming the CI fixes don't require major overhaul, <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r4.
Review status: 21 of 22 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[solvers/mixed_integer_rotation_constraint.h, line 24 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbIjWx-wjGpHqfO8pS:-LBc0vDn7VsSFcCQCtfA:bn9tz10) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L24)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. I added the formulation on the SO(3) constraint. Down below where I mention `IntervalBinning`, I explained how many binary variables will be introduced, by either linear or logarithmic binning.
</blockquote></details>

That helps a lot, thanks!

---

*[solvers/mixed_integer_rotation_constraint.h, line 58 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbKlDU9b60LaLF0SN9:-LBc1LeyCeg0Eo7U2rnr:b-uiz7lp) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.h#L58)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think the type of `B_` suggest that it is a 2D array of decision variable vectors? I added the comment that `B_ contains the new binary variables added to the program.`
</blockquote></details>

That's the main part I was concerned about, thanks!

---

*[solvers/mixed_integer_rotation_constraint.cc, line 928 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBbQwOb1X9Rbsz8UM-U:-LBc336998AGRf1g2Gbf:by2u7do) ([raw file](https://github.com/robotlocomotion/drake/blob/1ff39d15b64091f9e1621aa41497e701a4d9110b/solvers/mixed_integer_rotation_constraint.cc#L928)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

That is a good point! I agree it is more extensible, if we add more approaches. On the other hand, I have the concern that the user can misuse a wrong input, like `0b00` or `0b100`, as there is not a bitfield class in C++, while using enum class the user is guaranteed to use a right input.

Given this concern, do you know a good way in C++, to guarantee that the user chooses a valid input using bit field?
</blockquote></details>

Not that I'm aware of -- looks like std::bitset could make it impossible to submit exceedingly-long bit strings (though I don't know what kind of errors it would cause), but illegal combinations would have to be rejected as a pile of special cases.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBc4HKZA4i304vtibur:b-u1o5it)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Sherm, as I warned, this gigantic PR will be on your review stack +@sherm1 for platform review please.

---

Review status: 20 of 22 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBc8io13zK1vGFq4UQb:b-jyc092)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 20 of 22 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[solvers/test/mixed_integer_rotation_constraint_internal_test.cc, line 2 at r5](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBcG4NE03IwQzYKhS27:-LBcG4NE03IwQzYKhS28:b-9itgld) ([raw file](https://github.com/robotlocomotion/drake/blob/6dfcd04d1e0f822a3b7c3f1616acefb9f6411f14/solvers/test/mixed_integer_rotation_constraint_internal_test.cc#L2)):*
> ```cpp
> /* clang-format off to disable clang-format-includes */
> #include "drake/solvers/mixed_integer_rotation_constraint_internal.h"
> ```

To reviewers: this file is renamed from `rotation_constraint_internal_test.cc`, no new code is introduced. You may ignore reviewing it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending an assortment of BTWs.

---

Reviewed 7 of 23 files at r1.
Review status: all files reviewed at latest revision, 25 unresolved discussions.

---

*[multibody/global_inverse_kinematics.h, line 30 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhfjOF-9-x84V8BcVZ:-LBhfjOF-9-x84V8BcV_:byqbk8j) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/multibody/global_inverse_kinematics.h#L30)):*
> ```objc
>   struct Options {
>     // This constructor is needed, otherwise the compiler complains.
>     Options() {}
> ```

BTW is that still a problem? What error is produced?

---

*[solvers/mixed_integer_optimization_util.cc, line 21 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhgf_W6COEm0rvBaOY:-LBhgf_W6COEm0rvBaOZ:b-kzigak) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_optimization_util.cc#L21)):*
> ```cpp
>   // The following line should not be reached. We add it due to a compiler
>   // defect.
>   throw std::runtime_error("Should not reach this part of the code.\n");
> ```

BTW consider DRAKE_ABORT_MSG here -- something very bad must have happened if this gets executed!

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBhgrTG8ARkbgbmzdZ1:b8zzfnv)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Hongkai, if you haven't already please run kcov to verify that all the new code has unit tests.

---

Review status: all files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBhqmfHAy-vPNN_340d:b-e4w2bc)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Good call. I just ran kcov, and every line in `mixed_integer_rotation_constraint.cc` is covered.

@drake-jenkins-bot mac-highsierra-clang-bazel-experimental-everything please
@drake-jenkins-bot mac-sierra-clang-bazel-experimental-everything please
@drake-jenkins-bot mac-highsierra-clang-bazel-experimental-everything-debug please

---

Review status: 14 of 22 files reviewed at latest revision, 18 unresolved discussions.

---

*[multibody/global_inverse_kinematics.h, line 30 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhfjOF-9-x84V8BcVZ:-LBiiR1IChPmRDnsUpCb:b-dggywt) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/multibody/global_inverse_kinematics.h#L30)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW is that still a problem? What error is produced?
</blockquote></details>

It causes an error below in the constructor for `GlobalInverseKinematics`, when we construct the default options.
```
ERROR: /home/hongkai/drake-distro/multibody/BUILD.bazel:19:1: C++ compilation of rule '//multibody:global_inverse_kinematics' failed (Exit 1)
In file included from multibody/global_inverse_kinematics.cc:1:
bazel-out/k8-opt/bin/multibody/_virtual_includes/global_inverse_kinematics/drake/multibody/global_inverse_kinematics.h:52:61: error: default member initializer for 'num_intervals_per_half_axis_' needed within definition of enclosing class 'GlobalInverseKinematics' outside of member functions
                                   const Options& options = Options());
 ```

---

*[solvers/mixed_integer_optimization_util.cc, line 21 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhgf_W6COEm0rvBaOY:-LBij_EpCehtvjNXpdE9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_optimization_util.cc#L21)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider DRAKE_ABORT_MSG here -- something very bad must have happened if this gets executed!
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 37 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhQcoh15PPaLFniCbl:-LBhj9fc92zMo9TQUaHO:brny7ur) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L37)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW insert a space after last enumeration. (Otherwise this line gets sucked into point 3.)
</blockquote></details>

Done. Add an empty line here. Otherwise linting tools is unhappy with a trailing empty space.

---

*[solvers/mixed_integer_rotation_constraint.h, line 46 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhRYRlAPe7Ge2p45bH:-LBhjLa45MyXsjXFqKDu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L46)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW per f2f this should be just `AddToProgram()` (with parentheses). Doxygen will make it a link since the method is in the current class.
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 119 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhV72O7C6DriRZMHnC:-LBhjbdPDPkSNql0J3zv:b7yil97) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L119)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW normally output arguments should appear last in the argument list per GSG. Is there a contrary convention in MathProgram that the program comes first? If not please make this the last parameter.
</blockquote></details>

Done. I was following the original `AddRotationMatrixBilinearMccormickMilpConstraints`, which puts `prog` before `R`. But I guess it is better to follow GSG here.

---

*[solvers/mixed_integer_rotation_constraint.h, line 131 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhT6ibD__n_OtbJD6J:-LBhjxUH3ZVDql4zzDwr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L131)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI you don't have to document simple getters with obvious meaning.
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 172 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhTxGL2WWtJUV0-yMf:-LBhlHGd4f1EpJ-n1rRI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L172)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider making this a struct (like ReturnType above) -- tuples are awkward since the elements are unnamed. Here element pairs 0,1 and 2,3 have the same types so could be easily confused.
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 182 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhVTDS0J4XYUpJIMuX:-LBhlKrDB3vKTh6oDh-x:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L182)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW you need a blank line before this one or it sticks to the last bullet.
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 225 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhUhUb4q-gvPG-SExg:-LBhlMhtD_9yppl_Waui:b-g384zr) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L225)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW normally output arguments should appear last in the argument list per GSG. Is there a contrary convention in MathProgram that the program comes first? If not please make this the last parameter.
</blockquote></details>

Done. I copied this code directly from `rotation_constraint.h`, it was in the wrong order before.

---

*[solvers/mixed_integer_rotation_constraint.cc, line 351 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhW-LFD2Hprfd7IK6n:-LBhpCd-D90SDjrWMegY:bwxdexl) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.cc#L351)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW odd formatting
</blockquote></details>

Unfortunately that is what clang-format did automatically. I add clang-format off here.

---

*[solvers/mixed_integer_rotation_constraint.cc, line 855 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhWFIV9-WCKVRPwTnk:-LBhpc7I1t22xwxbe70a:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.cc#L855)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW initializer list mixes `{}` and `()` somewhat randomly -- please use them consistently.
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.cc, line 880 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhaK53EUWEMTWhmJ9C:-LBieyjoAjlS2encIoD1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.cc#L880)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider a comment before this line saying what the i,j loop is intended to do.
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.cc, line 910 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhaSK--BWMQTIZMSie:-LBih5TwAxQF4cwhzboU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.cc#L910)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW a comment here would be helpful, similar to the nice comment just below.
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.cc, line 1060 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhb0Ut17-p3T3nJGBh:-LBiiKIu8oCID0e9kLDf:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.cc#L1060)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW would be nice to see a struct here.
</blockquote></details>

Done.

---

*[solvers/rotation_constraint.cc, line 3 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhNeq49umJKRfwhBPk:-LBihJ0E43ufVHznQG4q:b-376vyo) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/rotation_constraint.cc#L3)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I don't think you need these clang-format guards any more since the internal header is gone.
</blockquote></details>

Done. Good catch!

---

*[solvers/test/mixed_integer_rotation_constraint_test.cc, line 36 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhc9qeB7JJtb5APkJG:-LBihTsdDB9i8snjPsu4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/test/mixed_integer_rotation_constraint_test.cc#L36)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW outer parentheses not need here.
</blockquote></details>

Done.

---

*[solvers/test/mixed_integer_rotation_constraint_test.cc, line 139 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhdGOdERyG2JXNkvfd:-LBihakbD_bCAirVd9V_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/test/mixed_integer_rotation_constraint_test.cc#L139)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider reflowing this comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBiqlDl0WrdA9l0XjiZ:b-e1p8a3)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 14 of 22 files reviewed at latest revision, 14 unresolved discussions.

---

*[solvers/test/mixed_integer_rotation_constraint_test.cc, line 223 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhe9ea9KEU6QIcTDyO:-LBit-sB-zHBMAbmAjyr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/test/mixed_integer_rotation_constraint_test.cc#L223)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW META should use `FAIL() << "message";` here, see [gtest doc](https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#explicit-success-and-failure). 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Very nice -- thanks! CI trouble?

---

Reviewed 8 of 9 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[multibody/global_inverse_kinematics.h, line 30 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhfjOF-9-x84V8BcVZ:-LBlWuY38iZzAZn1onva:b-ok7kbb) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/multibody/global_inverse_kinematics.h#L30)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

It causes an error below in the constructor for `GlobalInverseKinematics`, when we construct the default options.
```
ERROR: /home/hongkai/drake-distro/multibody/BUILD.bazel:19:1: C++ compilation of rule '//multibody:global_inverse_kinematics' failed (Exit 1)
In file included from multibody/global_inverse_kinematics.cc:1:
bazel-out/k8-opt/bin/multibody/_virtual_includes/global_inverse_kinematics/drake/multibody/global_inverse_kinematics.h:52:61: error: default member initializer for 'num_intervals_per_half_axis_' needed within definition of enclosing class 'GlobalInverseKinematics' outside of member functions
                                   const Options& options = Options());
 ```
</blockquote></details>

Oh yeah I remember now.

---

*[solvers/mixed_integer_optimization_util.cc, line 21 at r7](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBlX0on1ikuztZMXD-b:-LBlX0on1ikuztZMXD-c:bdxzure) ([raw file](https://github.com/robotlocomotion/drake/blob/08758eb68eb9012a51cd264ec844263de463c164/solvers/mixed_integer_optimization_util.cc#L21)):*
> ```cpp
>   // The following line should not be reached. We add it due to a compiler
>   // defect.
>   DRAKE_ABORT_MSG("Should not reach this part of the code.\n");
> ```

FYI the "\n" is not needed here (wasn't needed before either).

---

*[solvers/mixed_integer_rotation_constraint.h, line 37 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhQcoh15PPaLFniCbl:-LBlXLUx5ErdcQZjvgIL:byain9f) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L37)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. Add an empty line here. Otherwise linting tools is unhappy with a trailing empty space.
</blockquote></details>

Sorry, I meant a blank line rather than a space.

---

*[solvers/mixed_integer_rotation_constraint.cc, line 1060 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhb0Ut17-p3T3nJGBh:-LBlY16C20O_-GQqDrIn:b6xapi5) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.cc#L1060)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Much nicer, thanks!

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBlYTMg1UeMH2xPa6Tt:b-5fcg14)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Gurobi solver has different result on Mac machine versus Linux. Relaxed the tolerance a little bit.

@drake-jenkins-bot mac-highsierra-clang-bazel-experimental-everything please
@drake-jenkins-bot mac-sierra-clang-bazel-experimental-everything please
@drake-jenkins-bot mac-highsierra-clang-bazel-experimental-everything-debug please

---

Review status: 21 of 23 files reviewed at latest revision, all discussions resolved.

---

*[solvers/mixed_integer_optimization_util.cc, line 21 at r7](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBlX0on1ikuztZMXD-b:-LBnBDL04jPgS8LGfGlB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/08758eb68eb9012a51cd264ec844263de463c164/solvers/mixed_integer_optimization_util.cc#L21)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI the "\n" is not needed here (wasn't needed before either).
</blockquote></details>

Done.

---

*[solvers/mixed_integer_rotation_constraint.h, line 37 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8748#-LBhQcoh15PPaLFniCbl:-LBnBEoR-UeU-lYvWtXT:blyphgc) ([raw file](https://github.com/robotlocomotion/drake/blob/a035448c9c8e478129739a9afdfd6858c8ffdb09/solvers/mixed_integer_rotation_constraint.h#L37)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Sorry, I meant a blank line rather than a space.
</blockquote></details>

No problem!

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBnBXIQ4UAROHKG9n2P:bqrzb0i)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8748#-:-LBscq7YCILRZUuPastw:b-l70wzf)*
<!-- Sent from Reviewable.io -->
